packages/massa-web3/src/experimental/basicElements/address.ts

```diff
@@ -43,58 +45,102 @@ export class Address {
         throw new Error(`unsupported version: ${version}`)
     }
   }
 
+  private getPrefix(str: string): string {
+    const expected = [ADDRESS_USER_PREFIX, ADDRESS_CONTRACT_PREFIX]
+    for (let prefix of expected) {
+      if (str.startsWith(prefix)) {
+        return prefix
+      }
+    }
+    throw new Error(
+      `invalid address prefix: one of ${expected.join(' or ')} was expected.`
+    )
+  }
+
   /**
    * Initializes a new address object from a serialized string.
    *
    * @param str - The serialized address string.
-   * @param version - The version of the address. If not defined, the last version will be used.
    *
    * @returns A new address instance.
    *
    * @throws If the address string is invalid.
    */
-  public static fromString(str: string, version?: Version): Address {
+  public static fromString(str: string): Address {
+    const version = Address.getVersion(str)
     const address = Address.initFromVersion(version)
 
     try {
-      checkPrefix(
-        str,
-        ADDRESS_PREFIX + ADDRESS_USER_PREFIX,
-        ADDRESS_PREFIX + ADDRESS_CONTRACT_PREFIX
-      )
-      address.isEOA = str[ADDRESS_PREFIX.length] === ADDRESS_USER_PREFIX
-      address.bytes = extractData(
-        address.serializer,
-        address.versioner,
-        str.slice(ADDRESS_PREFIX.length + ADDRESS_USER_PREFIX.length),
-        address.version
+      const prefix = address.getPrefix(str)
+
+      address.isEOA = prefix === ADDRESS_USER_PREFIX
+      const versionedBytes = address.serializer.deserialize(
+        str.slice(prefix.length)
       )
+
+      address.bytes = Uint8Array.from([
+        ...varint.encode(address.isEOA ? 0 : 1),
+        ...versionedBytes,
+      ])
     } catch (e) {
       throw new Error(`invalid address string: ${e.message}`)
     }
-
     return address
   }
 
+  /**
+   * Get the address type from bytes.
+   *   *
+   * @returns the address type enum.
+   */
+  private getType(): number {
+    if (!this.bytes) {
+      throw new Error('address bytes is not initialized')
+    }
+    return varint.decode(this.bytes)
+  }
+
+  /**
+   * Get the address version.
+   *   *
+   * @returns the address type enum.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  static getVersion(data: string | Uint8Array): Version {
+    // when a new version will come, implement the logic to detect version here
+    // This should be done without serializer and versionner as they are potentially not known at this point
+    return Version.V0
+  }
+
   /**
    * Initializes a new address object from a public key.
    *
    * @param publicKey - The public key to derive the address from.
    *
    * @returns A new address object.
    */
-  public static fromPublicKey(publicKey: PublicKey): Address {
-    const address = Address.initFromVersion()
-    address.bytes = publicKey.hasher.hash(publicKey.toBytes())
+  public static fromPublicKey(
+    publicKey: PublicKey,
+    version = DEFAULT_VERSION
```

Comment:
Why do we need a version here?

Reply:
I think its the only case where we cannot deduce it ? and we need it to create address bytes

Reply:
I disagree. This implementation is only valid for a specific public key and address version.
My suggestion is to:
- remove the version argument,
- check that the publicKey version is V0 - otherwise throw an error,
- initFromVersion using V0

And keep the rest of the code.